backends/dpdk/dpdkPnaDirection.cpp

```diff
@@ -0,0 +1,124 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include "dpdkPnaDirection.h"
+#include "dpdkUtils.h"
+
+namespace DPDK {
+const IR::Node* AddNewMetadataFields::preorder(IR::DpdkStructType *st) {
+    if (isMetadataStruct(st)) {
+        for (auto nf : newMetadataFields) {
+                st->fields.push_back(nf);
+        }
+        auto newSt = new IR::DpdkStructType(st->srcInfo, st->name,
+                                          st->annotations, st->fields);
+        newMetadataFields.clear();
+        return newSt;
+    }
+    return st;
+}
+
+const IR::Node* DirectionToRegRead::preorder(IR::DpdkAsmProgram *p) {
+        p->externDeclarations.push_back(addRegDeclInstance(registerInstanceName));
+        return p;
+}
+
+IR::DpdkExternDeclaration* DirectionToRegRead::addRegDeclInstance(cstring instanceName) {
+    auto typepath = new IR::Path("Register");
+    auto type = new IR::Type_Name(typepath);
+    auto typeargs = new IR::Vector<IR::Type>({IR::Type::Bits::get(32),
+                                              IR::Type::Bits::get(32)});
+    auto spectype = new IR::Type_Specialized(type, typeargs);
+    auto args = new IR::Vector<IR::Argument>();
+    args->push_back(new IR::Argument(new IR::Constant(IR::Type::Bits::get(32), 1)));
+    auto annot = IR::Annotations::empty;
+    annot->addAnnotationIfNew(IR::Annotation::nameAnnotation,
+                              new IR::StringLiteral(instanceName));
+    auto decl = new IR::DpdkExternDeclaration(instanceName, annot, spectype, args,
+                    nullptr);
+    return decl;
+}
+
+void DirectionToRegRead::addMetadataField(cstring fieldName) {
+    if (newFieldName.count(fieldName))
+        return;
+     newMetadataFields.push_back(new IR::StructField(IR::ID(fieldName),
+                                 IR::Type::Bits::get(32)));
+     newFieldName.insert(fieldName);
+}
+
+bool DirectionToRegRead::isDirection(const IR::Member *m) {
+    if (m == nullptr)
+        return false;
+     return m->member.name == "pna_main_input_metadata_direction"
+         || m->member.name == "pna_pre_input_metadata_direction"
+         || m->member.name == "pna_main_parser_input_metadata_direction";
+}
+
+const IR::Node *DirectionToRegRead::postorder(IR::DpdkListStatement *l) {
+    l->statements = replaceDirectionWithRegRead(l->statements);
+    newStmts.clear();
+    return l;
+}
+
+const IR::Node *DirectionToRegRead::postorder(IR::DpdkAction *a) {
+    a->statements = replaceDirectionWithRegRead(a->statements);
+    newStmts.clear();
+    return a;
+}
+
+void DirectionToRegRead::replaceDirection(const IR::Member *m) {
+    addMetadataField(reg_read_tmp);
+    addMetadataField(left_shift_tmp);
+    auto reade = new IR::Member(new IR::PathExpression(IR::ID("m")),
```

Comment:
How are making sure that the name is unique?

Reply:
done